### PR TITLE
refactor: Migrate ZMCredentials from Objective-C to UserCredentials in Swift

### PR DESCRIPTION
**Pull request by agisilaos**
With this PR @KaterinaWire and I spent some time migrating `ZMCredentials` from Objective-C to `UserCredentials` in Swift. In addition to that we migrated `ZMPhoneCredentials` to `UserPhoneCredentials` and `ZMEmailCredentials` to `UserEmailCredentials`.

---

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

### CODE REVIEW

File: wire-ios-sync-engine/Source/UserSession/UserCredentials.swift [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    public var phoneNumberVerificationCode: String?
+    public var emailVerificationCode: String?
+
+    public init(
+        email: String? = nil,
+        password: String? = nil,
+        phoneNumber: String? = nil,
+        phoneNumberVerificationCode: String? = nil,
+        emailVerificationCode: String? = nil
+    ) {
```
Comment (markusfassbender) [inline wire-ios-sync-engine/Source/UserSession/UserCredentials.swift:35]: - Does it make sense to have a credential object where everything is `nil`?
- Is there a case where only the `password` and `phoneNumberVerificationCode` is set?

Maybe we can declarer the init a bit clearer or even seperate this in different objects as it seems to have multiple responsibilities for multiple tasks?

File: wire-ios-sync-engine/Source/UnauthenticatedSession/ZMAuthenticationStatus.swift [hunk 2]
```diff
@@ -19,3 +19,24 @@
...
+
+    public func emailCredentials() -> UserEmailCredentials! {
+        if let loginCredentials = self.loginCredentials, loginCredentials.credentialWithEmail {
+            return UserEmailCredentials.credentials(
+                email: loginCredentials.email ?? "",
+                password: loginCredentials.password ?? "",
+                emailVerificationCode: loginCredentials.emailVerificationCode
+            )
+        }
+        return nil
```
Comment (netbe) [inline wire-ios-sync-engine/Source/UnauthenticatedSession/ZMAuthenticationStatus.swift]: we return nil but this is explicitly unwrapped : `public func emailCredentials() -> UserEmailCredentials!`

File: wire-ios-sync-engine/Tests/Source/Synchronization/SynchronizationMocks.swift [hunk 3]
```diff
@@ -166,8 +166,8 @@ class FakeCredentialProvider: NSObject, ZMCredentialProvider {
     var email = "hello@example.com"
     var password = "verySafePassword"
 
-    func emailCredentials() -> ZMEmailCredentials! {
-        return ZMEmailCredentials(email: email, password: password)
+    func emailCredentials() -> UserEmailCredentials! {
```
Comment (netbe) [inline wire-ios-sync-engine/Tests/Source/Synchronization/SynchronizationMocks.swift]:
```suggestion
    func emailCredentials() -> UserEmailCredentials {
```
is that possible?

**Review by netbe — COMMENTED** [review]
add some comments about `!`

File: wire-ios-sync-engine/Source/UserSession/UserCredentials.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserCredentials.swift:35]: @markusfassbender phoneNumber and phoneNumberVerificationCode will be gone soon (I already opened a PR) so we're going to touch that file again. Do you think is needed for this PR?

Comment (markusfassbender) [reply wire-ios-sync-engine/Source/UserSession/UserCredentials.swift:35]: okay didn't know about it

**Review by markusfassbender — APPROVED** [review]

File: wire-ios-sync-engine/Tests/Source/Synchronization/SynchronizationMocks.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Tests/Source/Synchronization/SynchronizationMocks.swift]: Addressed it here: https://github.com/wireapp/wire-ios/commit/e93fc4bda3cdc961678f81f7980856357a3d3ef3

File: wire-ios-sync-engine/Source/UnauthenticatedSession/ZMAuthenticationStatus.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UnauthenticatedSession/ZMAuthenticationStatus.swift]: Addressed it here: https://github.com/wireapp/wire-ios/commit/e93fc4bda3cdc961678f81f7980856357a3d3ef3

**Review by netbe — APPROVED** [review]

**Pull request merged**